onnxruntime/core/mlas/lib/qgemm_kernel_neon.cpp

```diff
@@ -508,3 +496,477 @@ const MLAS_GEMM_U8X8_DISPATCH MlasGemmU8X8DispatchNeon = {
+            a3 += 16;
+
+            RowSums0 = vpadalq_s16(RowSums0, vpaddlq_s8(v0));
+            RowSums1 = vpadalq_s16(RowSums1, vpaddlq_s8(v1));
+            RowSums2 = vpadalq_s16(RowSums2, vpaddlq_s8(v2));
+            RowSums3 = vpadalq_s16(RowSums3, vpaddlq_s8(v3));
+
+            vst1q_u8(&D[0], vreinterpretq_u8_s8(v0));
+            vst1q_u8(&D[16], vreinterpretq_u8_s8(v1));
+            vst1q_u8(&D[32], vreinterpretq_u8_s8(v2));
+            vst1q_u8(&D[48], vreinterpretq_u8_s8(v3));
+
+            D += 64;
+            k -= 16;
+        }
+
+        if (k > 0) {
+
+            uint8_t* d = D;
+
+            vst1q_u8(&D[0], BitFlipVector);
+            vst1q_u8(&D[16], BitFlipVector);
+            vst1q_u8(&D[32], BitFlipVector);
+            vst1q_u8(&D[48], BitFlipVector);
+
+            if (k >= 8) {
+
+                vst1_u8(&d[0], vld1_u8(a0));
+                a0 += 8;
+                vst1_u8(&d[16], vld1_u8(a1));
+                a1 += 8;
+                vst1_u8(&d[32], vld1_u8(a2));
+                a2 += 8;
+                vst1_u8(&d[48], vld1_u8(a3));
+                a3 += 8;
+
+                d += 8;
+                k -= 8;
+            }
+
+            if (k >= 4) {
+
+                uint32_t v0 = *reinterpret_cast<const uint32_t*>(a0);
+                a0 += 4;
+                uint32_t v1 = *reinterpret_cast<const uint32_t*>(a1);
+                a1 += 4;
+                uint32_t v2 = *reinterpret_cast<const uint32_t*>(a2);
+                a2 += 4;
+                uint32_t v3 = *reinterpret_cast<const uint32_t*>(a3);
+                a3 += 4;
+
+                *reinterpret_cast<uint32_t*>(&d[0]) = v0;
+                *reinterpret_cast<uint32_t*>(&d[16]) = v1;
+                *reinterpret_cast<uint32_t*>(&d[32]) = v2;
+                *reinterpret_cast<uint32_t*>(&d[48]) = v3;
+
+                d += 4;
+                k -= 4;
+            }
+
+            while (k > 0) {
+
+                d[0] = *a0++;
+                d[16] = *a1++;
+                d[32] = *a2++;
+                d[48] = *a3++;
+
+                d += 1;
+                k -= 1;
+            }
+
+            int8x16_t v0 = vreinterpretq_s8_u8(veorq_u8(vld1q_u8(&D[0]), BitFlipVector));
+            int8x16_t v1 = vreinterpretq_s8_u8(veorq_u8(vld1q_u8(&D[16]), BitFlipVector));
+            int8x16_t v2 = vreinterpretq_s8_u8(veorq_u8(vld1q_u8(&D[32]), BitFlipVector));
+            int8x16_t v3 = vreinterpretq_s8_u8(veorq_u8(vld1q_u8(&D[48]), BitFlipVector));
+
+            RowSums0 = vpadalq_s16(RowSums0, vpaddlq_s8(v0));
+            RowSums1 = vpadalq_s16(RowSums1, vpaddlq_s8(v1));
+            RowSums2 = vpadalq_s16(RowSums2, vpaddlq_s8(v2));
+            RowSums3 = vpadalq_s16(RowSums3, vpaddlq_s8(v3));
+
+            vst1q_u8(&D[0], vreinterpretq_u8_s8(v0));
+            vst1q_u8(&D[16], vreinterpretq_u8_s8(v1));
+            vst1q_u8(&D[32], vreinterpretq_u8_s8(v2));
+            vst1q_u8(&D[48], vreinterpretq_u8_s8(v3));
```

Comment:
Intuitively, D is accessed multiple times. Is there a more efficient way?

Reply:
open to suggestions